File: TableViewReview/TableViewReview/CompinentTableView/OCComponent/OCComponentTableView/BaseAssemblyComponent.m
//
//  BaseAssemblyComponentViewController.m
//  TableViewReview
//
//  Created by Jialin Chen on 2019/7/31.
//  Copyright © 2019年 CJL. All rights reserved.
//

#import "BaseAssemblyComponent.h"
#import "BaseAssemblyComponent+Private.h"

@interface BaseAssemblyComponent ()

@end

@implementation BaseAssemblyComponent

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.
}

- (void)setupComponent{
    NSLog(@"setupComponent ! ");
}

- (BOOL)shouldShowComponent{
    return YES;
}

#pragma mark - Register Cell
- (void)registerClass:(Class)cellClass forCellReuseIdentifier:(NSString *)identifier{
    [self.tableView registerClass:cellClass forCellReuseIdentifier:identifier];
}

- (void)reloadTableView{
    [self.tableView reloadData];
}

- (void)reloadTableViewComponent{
    [self.componentSections enumerateObjectsUsingBlock:^(id  _Nonnull obj, NSUInteger idx, BOOL * _Nonnull stop) {
        NSArray *arr = (NSArray *)obj;
        BOOL isContain = [arr containsObject:self.componentType];
        if (isContain) {
            NSIndexSet *moduleSet = [[NSIndexSet alloc] initWithIndex:idx];
            [self.tableView reloadSections:moduleSet withRowAnimation:UITableViewRowAnimationNone];
            *stop = YES;
        }
    }];
}

- (void)reloadTableViewComponentWithRow:(NSInteger)row{
    [self.componentSections enumerateObjectsUsingBlock:^(id  _Nonnull obj, NSUInteger idx, BOOL * _Nonnull stop) {
        NSArray *arr = (NSArray *)obj;
        BOOL isContain = [arr containsObject:self.componentType];
        if (isContain) {
            NSIndexPath *componentIndexPath = [[NSIndexPath alloc] indexPathByAddingIndex:idx];
            [self.tableView reloadRowsAtIndexPaths: [NSArray arrayWithObjects:componentIndexPath, nil] withRowAnimation:UITableViewRowAnimationNone];
            *stop = YES;
        }
    }];
}

#pragma mark - life cycle
- (void)assemblyComponentWillAppear:(BOOL)animated{
    NSLog(@"assemblyComponentWillAppear ! ");
}

- (void)assemblyComponentWillDisappear:(BOOL)animated{
    NSLog(@"assemblyComponentWillDisappear ! ");
}

#pragma mark - TB

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return 0;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return 0;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    return [[UITableViewCell alloc] init];
}

- (void)collectionView:(UICollectionView *)collectionView didDeselectItemAtIndexPath:(NSIndexPath *)indexPath {
    NSLog(@"BaseAssemblyComponent is select:%ld", (long)indexPath.row);
}

- (CGFloat)tableView:(UITableView *)tableView heightForHeaderInSection:(NSInteger)section{
    return 0;
}

- (CGFloat)tableView:(UITableView *)tableView heightForFooterInSection:(NSInteger)section{
    return 0;
}

- (UIView *)tableView:(UITableView *)tableView viewForHeaderInSection:(NSInteger)section{
    return [[UIView alloc] init];
}

- (UIView *)tableView:(UITableView *)tableView viewForFooterInSection:(NSInteger)section{
    return [[UIView alloc] init];
}

#pragma mark - no-used method
- (void)tableView:(UITableView *)tableView accessoryButtonTappedForRowWithIndexPath:(NSIndexPath *)indexPath{
    
}

- (BOOL)tableView:(UITableView *)tableView canEditRowAtIndexPath:(NSIndexPath *)indexPath{
    return NO;
}

- (BOOL)tableView:(UITableView *)tableView canFocusRowAtIndexPath:(NSIndexPath *)indexPath{
    return NO;
}

- (BOOL)tableView:(UITableView *)tableView canMoveRowAtIndexPath:(NSIndexPath *)indexPath{
    return NO;
}

@end

File: TableViewReview/TableViewReview/CompinentTableView/SwiftComponent/SwiftComponent/BaseAssemblyComponent.swift
//
//  BaseAssemblyComponent.swift
//  TableViewReview
//
//  Created by Jialin Chen on 2019/8/1.
//  Copyright © 2019年 CJL. All rights reserved.
//

import UIKit

class BaseAssemblyComponent: UIViewController, UITableViewDelegate, UITableViewDataSource {

    override func viewDidLoad() {
        super.viewDidLoad()

        // Do any additional setup after loading the view.
    }
    

    //MARK:--------public
    //MARK:--------Register Cell
    public func shouldShowComponent()->Bool{
        return true
    }
    public func registerClass(_ cellClass: AnyClass, forCellReuseIdentifier identifier: String){
        self.tableView?.register(cellClass, forCellReuseIdentifier: identifier)
    }
    public func reloadTableView(){
        self.tableView?.reloadData()
    }
    public func reloadTableViewComponent(){
        self.componentSections?.enumerateObjects({ (obj, idx, stop) in
            let arr = obj as! NSArray
            let isContain = arr.contains(self.componentType as Any)
            if isContain {
                let moduleSet = IndexSet.init(integer: idx)
                self.tableView?.reloadSections(moduleSet, with: UITableView.RowAnimation.none)
                stop[0] = true
            }
        })
    }
    public func reloadTableViewComponentWithRow(_ row: Int){
        self.componentSections?.enumerateObjects({ (obj, idx, stop) in
            let arr = obj as! NSArray
            let isContain = arr.contains(self.componentType as Any)
            if isContain {
                let componentIndexPath = IndexPath.init(row: row, section: idx)
                self.tableView?.reloadRows(at: [componentIndexPath], with: UITableView.RowAnimation.none)
                stop[0] = true
            }
        })
    }
    
    //MARK:--------Life Cycle
    public func setupComponent(){
        print("setupComponent !")
    }
    public func assemblyComponentWillAppear(_ animated: Bool){
        print("assemblyComponentWillAppear !")
    }
    public func assemblyComponentWillDisappear(_ animated: Bool){
        print("assemblyComponentWillDisappear !")
    }
    
    //MARK:--------private
    //MARK:--------TB
    func numberOfSections(in tableView: UITableView) -> Int {
        return 0
    }
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return 0
    }
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        return UITableViewCell.init()
    }
    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        print("BaseAssemblyComponent is select: \(indexPath.row)")
    }
    func tableView(_ tableView: UITableView, heightForHeaderInSection section: Int) -> CGFloat {
        return 0
    }
    func tableView(_ tableView: UITableView, heightForFooterInSection section: Int) -> CGFloat {
        return 0
    }
    func tableView(_ tableView: UITableView, viewForHeaderInSection section: Int) -> UIView? {
        return UIView.init()
    }
    func tableView(_ tableView: UITableView, viewForFooterInSection section: Int) -> UIView? {
        return UIView.init()
    }
    //MARK:--------no used method
    func tableView(_ tableView: UITableView, accessoryButtonTappedForRowWith indexPath: IndexPath) {
        
    }
    func tableView(_ tableView: UITableView, canEditRowAt indexPath: IndexPath) -> Bool {
        return false
    }
    func tableView(_ tableView: UITableView, canFocusRowAt indexPath: IndexPath) -> Bool {
        return false
    }
    func tableView(_ tableView: UITableView, canMoveRowAt indexPath: IndexPath) -> Bool {
        return false
    }
    
}
